[RFC] cmake: Add Cmake function for adding named sections to the linker file

oyvindronningstad opened the pull request:
Call add_named_section(".foo") to add a FLASH section with the name .foo.
Will also create the symbols _foo_start, _foo_end, and _foo_size.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad commented:
cc @hakonfam @sigvartmh @osaether

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Gitlint issues

3: UC4 Line exceeds max length (73>72): "Call add_named_section(".foo") to add a FLASH section with the name .foo."

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -689,6 +689,80 @@ function(zephyr_check_compiler_flag lang option check)
+# and _bar_size.
+#
+set(generated_sections_filename "${PROJECT_BINARY_DIR}/include/generated/generated-sections.ld")
+file(WRITE ${generated_sections_filename} "")
```
So far, unless I've missed a patch, we have been able to limit this file to functions and macros. I'd like this to continue to hold if possible.

oyvindronningstad replied:
Done

SebastianBoe commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -515,6 +515,10 @@ SECTIONS
+/* Located in generated directory. This file is populated by the
+ * add_named_section() Cmake function. */
+#include <generated-sections.ld>
+
```
We need to add this to all linker.ld unfortunately.

oyvindronningstad replied:
Will do

oyvindronningstad commented:
@SebastianBoe Thanks, I started addressing your comments. I ported CCM_*, TI_CCFG, and ramfunc so far. Take a look to see whether you think it's going in the right direction.

SebastianBoe commented:
Can you elaborate on the motivation?

Is it because we think it is easier for users to interface with the add_named_section API than the LD API, or is it to support out-of-tree use-cases?

oyvindronningstad commented:
> Can you elaborate on the motivation?

Yep, these are my main points:
 - Keep the linker definitions of a module together with everything else to do with that module.
 - Make it easier to add a section, syntax-wise.
 - Make it easier to enforce "best-practices", f.ex. the restriction on usage of `SECTION_DATA_PROLOGUE`
 - Make it less invasive to add a section, since you're less likely to interfere with what is already there. Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
 - Make the linker script cleaner and more readable.

The out-of-tree point is a bonus, hadn't thought of that :).

sigvartmh commented:
I really like this change as it cleans up the vendor specific stuff out of the linkerscript where it doesn't really belong.

SebastianBoe commented:
I'll get back to this after multi-image is merged.

There are definitely some good ideas here that we want to take advantage of :)

SebastianBoe commented on soc/arm/st_stm32/common/CMakeLists.txt:
```diff
@@ -1,2 +1,8 @@
 zephyr_sources(stm32cube_hal.c)
+
+if (DEFINED DT_CCM_BASE_ADDRESS)
+  add_named_section(".ccm_bss" REGION CCM GROUP_START CCM)
```
Are you missing NOLOAD?

SebastianBoe commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -472,45 +472,6 @@ SECTIONS
-	__ccm_data_end = .;
-	__ccm_end = .;
-
-	__ccm_data_rom_start = LOADADDR(_CCM_DATA_SECTION_NAME);
```
Are we missing this symbol?

SebastianBoe commented:
When I read this I am finding it easier to predict the behaviour of the original LD scripts than the new API.

My suggestion would therefore be to keep these improvements:

> Keep the linker definitions of a module together with everything else to do with that module.
> 
> Make it less invasive to add a section, since you're less likely to interfere with what is already there.
> 
> Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
> 
> Make the linker script cleaner and more readable.
> 
> Support injecting multiple out-of-tree LD script snippets.

And leave these be as-is:

> Make it easier to add a section, syntax-wise.
> 
> Make it easier to enforce "best-practices", f.ex. the restriction on usage of SECTION_DATA_PROLOGUE
> 

by having the API append/concatenate local LD script snippets instead of generating LD script itself.

oyvindronningstad commented:
> When I read this I am finding it easier to predict the behaviour of the original LD scripts than the new API.
> 
> My suggestion would therefore be to keep these improvements:
> 
> > Keep the linker definitions of a module together with everything else to do with that module.
> > Make it less invasive to add a section, since you're less likely to interfere with what is already there.
> > Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
> > Make the linker script cleaner and more readable.
> > Support injecting multiple out-of-tree LD script snippets.
> 
> And leave these be as-is:
> 
> > Make it easier to add a section, syntax-wise.
> > Make it easier to enforce "best-practices", f.ex. the restriction on usage of SECTION_DATA_PROLOGUE
> 
> by having the API append/concatenate local LD script snippets instead of generating LD script itself.

I think this is a good call. I opened a separate pull request for the updated work (#109)
